README.md
[![Build Status](https://travis-ci.org/rbarton65/espnff.svg?branch=master)](https://travis-ci.org/rbarton65/espnff) [![version](https://img.shields.io/badge/version-1.3.1-blue.svg)](https://github.com/rbarton65/espnff/blob/master/CHANGELOG.md) [![PyPI version](https://badge.fury.io/py/espnff.svg)](https://badge.fury.io/py/espnff)

# ESPN Fantasy Football API

Using ESPN's Fantasy Football private API, this package interfaces with 
ESPN Fantasy Football to gather data from any public league. A good way to mine data
without webscraping for data manipulation projects.

## Getting Started

These instructions will get you a copy of the project up and running 
on your local machine for development and testing purposes.

### Installing
With pip:

```python3
pip3 install espnff
```

With Git:

```bash
git clone https://github.com/rbarton65/espnff

cd espnff

python3 setup.py install
```


## Basic Usage

This gives an overview of all the features of `espnff`

### Downloading a public league

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league
League 123456, 2016 Season
```

### Viewing teams in a public league

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league.teams
[Team(Team 1), Team(Team 2), Team(Team 3), Team(Team 4),
Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8)]
>>> team1 = league.teams[0]
>>> team1
Team(Team 1)
```

### Viewing data for specific team

```python3
>>> league.teams
[Team(Team 1), Team(Team 2), Team(Team 3), Team(Team 4),
Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8)]
>>> team1 = league.teams[0]
>>> team1.team_id
1
>>> team1.team_name
Team 1
>>> team1.team_abbrev
T1
>>> team1.owner
Roger Goodell
>>> team1.division_id
0
>>> team1.division_name
Division 1
>>> team1.wins
5
>>> team1.losses
1
>>> team1.points_for
734.69
>>> team1.points_against
561.15
>>> team1.schedule
[Team(Team 2), Team(Team 3), Team(Team 4), Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8),
Team(Team 2), Team(Team 3), Team(Team 4), Team(Team 5), Team(Team 6), Team(Team 7), Team(Team 8)
>>> team1.scores
[135.5, 126.38, 129.53, 126.65, 114.81, 101.82, 1.15, 0, 0, 0, 0, 0, 0, 0]
>>> team1.mov
[32.12, 24.92, 45.97, 34.17, 41.74, -5.39, 1.15, 0, 0, 0, 0, 0, 0, 0]
```

### Viewing league settings

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> settings = league.settings
>>> settings
'Settings(League Name)'
>>> settings.reg_season_count
14
>>> self.final_season_count
16
>>> settings.undroppable_list
true
>>> settings.veto_votes_required
4
>>> settings.team_count
8
>>> settings.playoff_team_count
4
>>> settings.id
123456
>>> settings.keeper_count
3
>>> settings.tie_rule
'Most Bench Points'
>>> settings.playoff_seed_tie_rule
'Head to Head Record'
>>> settings.roster
{'RB/WR/TE': 1, 'BE': 7, 'QB': 1, 'D/ST': 1, 'RB': 2, 'TE': 1, 'K': 1, 'WR': 2}
>>> settings.trade_deadline
2016-11-16T17:00:00.000Z
>>> settings.name
League Name
>>> settings.status
playoffs
>>> settings.year
2016
>>> settings.server_date
2016-12-08T21:06:53.087Z
```

### Viewing power rankings

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league.power_rankings(week=5)
[('31.85', Team(Team 1)), ('25.60', Team(Team 3)), ('25.60', Team(Team 6)), ('22.45', Team(Team 2)),
('20.70', Team(Team 8)), ('18.20', Team(Team 7)), ('18.20', Team(Team 4)), ('18.10', Team(Team 5))]
```

### Viewing scoreboard

```python3
>>> from espnff import League
>>> league_id = 123456
>>> year = 2016
>>> league = League(league_id, year)
>>> league.scoreboard() # grab current week
["Matchup(Team(Team 2), Team(Team 7))", "Matchup(Team(Team 1), Team(Team 11))",
"Matchup(Team(Team 6), Team(Team 9))", "Matchup(Team(Team 12), Team(Team 4))",
"Matchup(Team(Team 10), Team(Team 3))", "Matchup(Team(Team 8), Team(Team 5))"]
>>> scoreboard = league.scoreboard(week=12) # define week
>>> scoreboard
["Matchup(Team(Team 2), Team(Team 7))", "Matchup(Team(Team 1), Team(Team 11))",
"Matchup(Team(Team 6), Team(Team 9))", "Matchup(Team(Team 12), Team(Team 4))",
"Matchup(Team(Team 10), Team(Team 3))", "Matchup(Team(Team 8), Team(Team 5))"]
>>> matchup = scoreboard[1]
>>> matchup
"Matchup(Team(Team 1), Team(Team 11))"
>>> matchup.home_team
"Team(Team 1)"
>>> matchup.home_score
7.05
>>> matchup.away_team
"Team(Team 11)"
>>> matchup.away_score
45.85
```

## Running the tests

Automated tests for this package are included in the `tests` directory. After installation,
you can run these tests by changing the directory to the `espnff` directory and running the following:

```python3
python3 setup.py test
```

## Versioning

This library uses [SemVer](http://semver.org/) for versioning. For available versions, see the
[tags on this repository](https://github.com/rbarton65/espnff/tags)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [Unreleased]
### Patched
- Fixed output for possibility the attribute doesn't exist in the API

### Other
- Reorganized structure of `espnff` package and tests for future development

## [1.3.1] - 2017-07-24
### Added
- Added CI testing for Python 3.6

### Patched
- Fixed issue where map for `playoff_seed_tie_rule` was wrong.

## [1.3.0] - 2016-12-12
### Added
- Added `settings` attribute for league instance.
- Added `roster`, `tie_rule`, and `playoff_seed_tie_rule` to `settings`

## [1.2.3] - 2016-12-06
### Patched
- Fixed issue where bye weeks were breaking `scoreboard()` function.

## [1.2.2] - 2016-11-30
### Patched
- Fixed `power_rankings()` issue where execution would print matrices.

## [1.2.1] - 2016-11-29
### Patched
- Fixed `power_rankings()` issue where lists were out of range.

## [1.2.0] - 2016-11-25
### Added
- Added scoreboard function to League instance.
- Added tests for scoreboard function.

### Removed
- Removed deprecation warnings, fully converted to new attributes.
- Removed need for `numpy` package.


## [1.1.0] - 2016-10-18
### Added
- `.travis.yml` file for CI testing.
- `setup.cfg` file for `flake8` testing.
- Http requests mock for testing.
- `league.status_code` for testing purposes.
- Exceptions for status codes.
- Tests for future development.

### Changed
- Deprecation warnings for `Team` class attributes. Next Version
will remove old attributes.
- README.md shows build status from Travis CI.
- Better documentation in README.md.
- Added range for dependencies.
- `from espnff import *` now only imports `League` and `Team`.


## [1.0.1] - 2016-10-18
### Added
- Tests for utility functions.

### Changed
- Output for matrix functions now lists instead of numpy array.

### Fixed
- Fix issue where not all teams showed in power rankings.
- Fix issue where power rankings output was not sorted.

## [1.0.0] - 2016-10-04
### Added
- New attributes for Teams.
- Started using Semantic Versioning.

### Changed
- Replaced urls with ESPN's private API url.
- Replaced `Members` class with `Teams`.
- `get_week` function now `power_rankings`
- Teams now accessable as list attribute in `League`.

### Removed
- `get_member`
- `get_all_members`

## 0.2.2 - 2016-09-27
### Fixed
- Fix issue where margin of victory calculation in the power rankings
algorithm accounted for 50% of the total power points. Should have been 5%.

## 0.2.1 - 2016-09-15
### Fixed
- Fix issue where scores from games that weren't played would become
a string character instead of an integer.

## 0.2.0 - 2016-07-03
### Changed
- Rearranged output for `get_week`.

## 0.1.0 - 2016-05-13
### Added
- Power rankings function `get_week` to `League` class.
- README information.
- `setup.py` to turn project into module.

### Changed
- Renamed `power_rankings.py` to `espnff.py`.

## 0.0.0 - 2016-05-12
### Added
- League and Members classes to obtain information about ESPN league.


[Unreleased]: https://github.com/rbarton65/espnff/compare/v1.3.1...HEAD
[1.3.1]: https://github.com/rbarton65/espnff/compare/v1.3.0...v1.3.1
[1.3.0]: https://github.com/rbarton65/espnff/compare/v1.2.3...v1.3.0
[1.2.3]: https://github.com/rbarton65/espnff/compare/v1.2.2...v1.2.3
[1.2.2]: https://github.com/rbarton65/espnff/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/rbarton65/espnff/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/rbarton65/espnff/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/rbarton65/espnff/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/rbarton65/espnff/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/rbarton65/espnff/releases/tag/v1.0.0


espnff/power_rankings.py
import re
import operator

import json
from lxml import html

import requests

from .utils import two_step_dominance


class League(object):

    def __init__(self, league_id, year):
        self.league_id = league_id
        self.year = year
        self.members = []
        self.basic_settings = {
                               'League Name': None,
                               'Number of Teams': None,
                               'Scoring Type': None
                               }
        self._fetch_members()

    def _fetch_members(self):
        '''Fetch members in league'''
        url = 'http://games.espn.go.com/ffl/standings?leagueId=%s&seasonId=%s'
        page = requests.get(url % (self.league_id, self.year))
        tree = html.fromstring(page.content)

        for member in tree.xpath('//tr[@class="tableBody"]//td//a'):
            team_name = member.text
            # parse team id from href url
            keysearch = re.compile('teamId=(.*?)&')
            team_id = keysearch.search(member.attrib['href']).group(1)
            self.members.append(Members(team_name, team_id, self.league_id, self.year))

    def _fetch_settings(self):
        '''Fetch settings for league'''
        url = 'http://games.espn.go.com/ffl/leaguesetup/settings?leagueId=%s&seasonId=%s'
        page = requests.get(url % (self.league_id, self.year))
        tree = html.fromstring(page.content)

    def _previous_ranking(self, week):
        '''Get last week power rankings for specified week'''
        week = week - 1
        wins_matrix = [x._calculate_wins(week) for x in self.members]
        # calculate two step dominance
        dominance_list = two_step_dominance(wins_matrix)

        # assign dominance number to each team, calculate power points
        for dom, team in zip(dominance_list, self.members):
            team.dominance = dom
            team._power_points(week)

        # sort by power points
        self.members.sort(key=operator.attrgetter('power_points'), reverse=True)

        for rank, team in enumerate(self.members):
            team.previous_rank = rank+1

    def _current_ranking(self, week):
        '''Get curent power rankings for specified week'''
        wins_matrix = [x._calculate_wins(week) for x in self.members]
        # calculate two step dominance
        dominance_list = two_step_dominance(wins_matrix)

        # assign dominance number to each team, calculate power points
        for dom, team in zip(dominance_list, self.members):
            team.dominance = dom
            team._power_points(week)

        # sort by power points
        self.members.sort(key=operator.attrgetter('power_points'), reverse=True)

        for rank, team in enumerate(self.members):
            team.current_rank = rank+1

    def get_week(self, week):
        '''Get power rankings for specified week'''
        if week > 1:
            self._previous_ranking(week)
            self._current_ranking(week)
        else:
            self._current_ranking(week)

        # build dict object
        data = {int(x.current_rank): {
                        'name': x.name,
                        'losses': x.losses,
                        'power points': x.power_points,
                        'previous rank': x.previous_rank,
                        'wins': x.wins
                        }
                for x in self.members
                }
        return data

    def get_member(self, id):
        '''Get team information'''
        # find team with matching team id
        data = {int(x.team_id): {
                            'name': x.name,
                            'year': x.year,
                            'scores': x.scores,
                            'opponents': x.opponents,
                            'wins': x.wins,
                            'mov': x.mov
                            }
                for x in self.members if x.team_id == str(id)
                }
        return data

    def get_all_members(self):
        '''Get all member information'''
        data = {int(x.team_id): {
                            'name': x.name,
                            'year': x.year,
                            'scores': x.scores,
                            'opponents': x.opponents,
                            'wins': x.wins,
                            'mov': x.mov
                            }
                for x in self.members
                }
        return data


class Members(object):

    def __init__(self, name, team_id, league_id, year):
        self.name = name
        self.team_id = team_id
        self.league_id = league_id
        self.year = year
        self.current_rank = 0
        self.losses = 0
        self.mov = []
        self.opponents = []
        self.previous_rank = 0
        self.power_points = 0
        self.scores = []
        self.wins = 0
        self.dominance = 0
        self._fetch_info()

    def _fetch_info(self):
        '''Fetch the scores, opponents, and margin of victory for each week'''
        url = 'http://games.espn.go.com/ffl/schedule?leagueId=%s&teamId=%s&seasonId=%s'
        page = requests.get(url % (self.league_id, self.team_id, self.year))
        tree = html.fromstring(page.content)

        # Fetch scores for regular season (13 weeks)
        for score in tree.xpath('//table//tr[@bgcolor]//td//nobr//a')[:13]:
            try:
                self.scores.append(float(score.text[2:].split('-')[0]))  # 'W 98-85' will return '98'
            except:
                self.scores.append(float(0))

        # Fetch opponents for regular season
        for opp in tree.xpath('//table//tr[@bgcolor]//td[@nobr=""]//a')[:13]:
            keysearch = re.compile('teamId=(.*?)&')
            self.opponents.append(keysearch.search(opp.attrib['href']).group(1))

        # Fetch margin of victory for regular season
        for score in tree.xpath('//table//tr[@bgcolor]//td//nobr//a')[:13]:
            if 'pts' not in score.text:  # 'pts' appear on bye weeks
                try:
                    team_score = float(score.text[2:].split('-')[0])  # 'W 98-85' will return '98', need float for decimals
                    opp_score = float(score.text.split('-')[1])  # 'W 98-89' will return '85', need float for decimals
                except:
                    # Game hasn't happened yet
                    team_score = opp_score = 0
                mov = team_score - opp_score
                self.mov.append('{0:.2f}'.format(mov))
                if mov > 0:
                    self.wins += 1

    def _power_points(self, week):
        '''Calculate power points'''
        # turn needed values into floats
        self.scores = [float(x) for x in self.scores]
        self.mov = [float(x) for x in self.mov]
        self.dominance = float(self.dominance)
        score_avg = float('{0:.2f}'.format(sum(self.scores[:week]) / week))
        mov_avg = float('{0:.2f}'.format(sum(self.mov[:week]) / week))
        # formula is weighted 80% dominance, 15% scores, 5% margin of victory
        self.power_points = float('{0:.2f}'.format((self.dominance*0.8) +
                                                   (score_avg*0.15) +
                                                   (mov_avg*0.05)))

    def _calculate_wins(self, week):
        '''Calculates wins based on negative or positive margin of victory'''
        wins = [0]*32  # team_id numbers increase with new members, 32 to be safe
        # create tuple to determine outcome against opponents
        outcomes = list(zip(self.opponents, self.mov))
        # convert to ints and floats
        outcomes = [(int(x), float(y)) for (x, y) in outcomes]

        for opp, mov in outcomes[:week]:
            opp = opp-1  # team id 1 is 0th place in self.wins
            if mov > 0:
                wins[opp] += 1

        # set losses and wins
        self.losses = week - sum(wins)
        self.wins = sum(wins)

        return wins
